### DISCUSSION TITLE
Support return if

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#4711] [Language Ideas]
Whenever I am to write

if(condition)
   return;
   
I find it more logical to be written

return if (condition);

The reason is that return statement has a more profound effect on the flow than if statement. Normal return, of course, would remain as such. Notice that in other places like where, the position of the word is more logical. 

break and continue are on a little bit different level since they break the flow of the local loop and I do not see it as severe as in case of return, but yes the same feature can be added there.

The statement if/return is written way too often as to not to be considered as a general improvement.

**Comment by Joe4evr** [comment]
#138, #325, #1368, #2966, #3242.

**Reply by ghost** [reply]
I disagree with all the above proposals because they might have some benefits but they are adding complexity.

My proposal is only about the order of return and if, really no more than that.

The main reason is that if done in this manner it becomes more documented what function can return where and the condition for it. This would encourage not to place return to deep and so on.

There would be no else statement, just if

return if (condition);
return if (condition) 25;

else if added would still mean return something

return if (condition) 25; else 30;

return 25; or otherwise return 30;

**Reply by cheong00** [reply]
IMO, the case of "return if (condition) 25; else 30;" do not need special handling because if you're going to exit the function anyway, the currently supported syntax of "return (condition) ? 25 : 30;" is equally readable, I think.

**Comment by HaloFour** [comment]
> The reason is that return statement has a more profound effect on the flow than if statement

IMO that's why there shouldn't be a conditional version of it.  When you see `return` at the beginning of a statement it should always result in the control flow leaving the current method.

**Comment by ghost** [comment]
Yeah, except that it is always under the condition that is stated above
that. So it is an implicit return anyway.

You are talking about the current agreement about return and take that as
something that is written in stone for logical reasons.
It is not. It is just picked up from C where it was unnecessary to have
such statement, since the intention was to keep the language reduced to few
most necessary statements.

If C# would remain the same and keep it low-key, I would agree. But C# is
going away from that intention, adding syntax sugars all around the place.

If so "return if" is more logical than "if return" since it clearly states
the intention which is "return" under the condition. Stating "if
(condition) return" is not different than changing the order of condition
and if statement:

(condtion)if

You say the condition first and then your intention - if. We do not do
that, right? We say

if(condition)

since the intention if, which is "to fork", goes upfront.

Even if not adopted

return if

is that many times more logical than "if(condition) return". Period.

**Reply by HaloFour** [reply]
> If so "return if" is more logical than "if return" since it clearly states
> the intention which is "return" under the condition.

"Under the condition" being the operative phrase there.  Your intention is to "fork" first, then to "return" from the method.  Your eye should be first drawn to the fact that the statement(s) will only happen as a result of the conditional.  As such, the current arrangement of the clauses is appropriate and correct.

**Comment by huoyaoyuan** [comment]
> I find it more logical to be written

It may depend on your mother language.
As a Chinese speaker, I always treat it more logical to put condition at first.

**Reply by cheong00** [reply]
For English, I don't have particular preference on "If you see fire on exit, return immediately" or "Return immediately if you see fire on exit".

**Comment by CyrusNajmabadi** [comment]
In C# we are generally amenable to having more than one way to do things, if we feel teh alternatives are substantively better/shorter enough to be meaningfully different.  In this case though, i don't see that.  The forms are virtually identical with exactly the same tokens/length, just reordered.  i.e.:

```c#
if(condition) return;
return if(condition);
```

So tehre is no real gain here as far as i can tell.  Just one more way to do the same thing, with no benefits over the existing form.

**Reply by proteo5** [reply]
I initiate a discussion about this same thing on #7927.
I can understand that this  `return (condition) value;` doesn't seems a lot of improvement over the regular `if (condition) return value;` 

But it seems that for a lot of us, the developers on the field, are asking for this since for us it really matter to do a thing like this.

Can you please reconsider?

**Reply by CyrusNajmabadi** [reply]
@proteo5 We regularly assess and consider all lang requests made.  :)  Thanks!
